### PR TITLE
Add SBOM generation

### PR DESCRIPTION
**Pull request by WilliamXieMSFT**
###### Summary 

This change supports generating the Software Bill of Materials (SBOM) on internal builds.

###### Change Log  

- Add PostBuild stage after Build stage that pulls down the Windows-Binaries-Release artifact to generate SBOM
- Upload SBOM as a build artifact

###### Test Methodology

Internal Build - https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5827380&view=results

### CODE REVIEW

**Comment by WilliamXieMSFT** [comment]
(Relevant to Internal only) Due to our setup with VS we don't actually need these SBOMs; our nuget packages get ingested by the core vsmanifest which generates an SBOM on our behalf. The SBOM generated by these changes is mainly for audit purposes.

File: build/yaml/pipelines/signed.yaml [hunk 1]
```diff
@@ -10,6 +10,7 @@ name: $(date:yyyyMMdd)$(rev:rr)
 
 variables:
   TeamName: ClrInstrumentationEngine
+  Packaging.EnableSBOMSigning: true
```
Comment (jander-msft) [inline build/yaml/pipelines/signed.yaml:13]: Should this just be a variable on the postbuild job instead of all of the jobs in this stage?

File: build/yaml/jobs/windows/postbuild.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# Job Template: PostBuild steps
+
+parameters:
+  SignType: None
```
Comment (jander-msft) [inline build/yaml/jobs/windows/postbuild.yaml:7]: Not used in this job.

File: build/yaml/pipelines/signed.yaml [hunk 3]
```diff
@@ -22,11 +23,17 @@ stages:
       Configuration: Release
       IsMicroBuildInternal: true
 
+- stage: PostBuild
+  jobs:
+  - template: ../jobs/windows/postbuild.yaml 
+    parameters:
```
Comment (jander-msft) [inline build/yaml/pipelines/signed.yaml:29]: None of these parameters are part of the postbuild.yaml template except SignType (which isn't used).

Comment (jander-msft) [reply build/yaml/pipelines/signed.yaml:29]: Maybe these were accidentally moved from the packages.yaml template inclusion?

File: build/yaml/steps/microbuild/codesignverify.yaml [hunk 4]
```diff
@@ -4,7 +4,8 @@
 # Step Template: MicroBuild Signing and Verification
 
 parameters:
-    SignType: None
+  SignType: None
+  TargetFolder: $(Build.StagingDirectory)
```
Comment (jander-msft) [inline build/yaml/steps/microbuild/codesignverify.yaml:8]: I don't see this parameter being used, Maybe revert this change.

File: build/yaml/jobs/windows/postbuild.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+jobs:
+- job: PostBuild
```
Comment (jander-msft) [inline build/yaml/jobs/windows/postbuild.yaml:10]: My rough reading of the requirements for SBOM generation is that this has to contain everything that is publicly shipped. Arguably, this would include all packages whether they are NuGet, MSI, MSM, etc. This probably should be a post-package job and may only need to consider the packages themselves rather than all binaries; the requirements state that only top-level dependencies are required.

**Review by jander-msft — CHANGES_REQUESTED** [review]

File: build/yaml/pipelines/signed.yaml (hunk 3 above)
Comment (WilliamXieMSFT) [reply build/yaml/pipelines/signed.yaml:29]: Oops, fixing

File: build/yaml/jobs/windows/postbuild.yaml (hunk 2 above)
Comment (WilliamXieMSFT) [reply build/yaml/jobs/windows/postbuild.yaml:7]: Leftover artifact, removing

File: build/yaml/jobs/windows/postbuild.yaml (hunk 5 above)
Comment (WilliamXieMSFT) [reply build/yaml/jobs/windows/postbuild.yaml:10]: From the guidance:
```
You should run the generator after you have built your binaries but before you have built your installer manifest and setup packages. The SBOM generator task takes an output location (BuildDropPath) as input and generates an SBOM file for that output. The output directory you specify should contain every file contained the in the packages you are planning on inserting into Visual Studio. 
```

I'll refactor this to pull in only the files we put into our packages.

Comment (jander-msft) [reply build/yaml/jobs/windows/postbuild.yaml:10]: Other guidance that I've seen is `it is okay to generate an SBOM that just lists the NuGet package and not the contents.`, which tells me that the package is what's important, not the individual files within the package. I've sent some email out to hopefully get some clarification on this.

File: build/yaml/jobs/windows/postbuild.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  displayName: PostBuild
+
+  variables:
+    Packaging.EnableSBOMSigning: true
+
+  pool:
+    name: VSEngSS-MicroBuild2019-1ES
+    demands:
```
Comment (jander-msft) [inline build/yaml/jobs/windows/postbuild.yaml:15]: I don't think any of these demands are necessary.

File: build/yaml/jobs/windows/postbuild.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    clean: outputs
+
+  steps:
+  - checkout: none
+
+  - template: ../../steps/azuredevops/downloadbuildartifacts.yaml
+    parameters:
+      ArtifactName: binaries-windows-Release
+
+  - template: ../../steps/microbuild/sbom.yaml
```
Comment (jander-msft) [inline build/yaml/jobs/windows/postbuild.yaml:29]: We need more than just the Windows binaries; we definitely need the Linux binaries.

**Review by jander-msft — CHANGES_REQUESTED** [review]

File: build/yaml/jobs/windows/postbuild.yaml (hunk 6 above)
Comment (WilliamXieMSFT) [reply build/yaml/jobs/windows/postbuild.yaml:15]: removing

File: build/yaml/jobs/windows/postbuild.yaml (hunk 7 above)
Comment (WilliamXieMSFT) [reply build/yaml/jobs/windows/postbuild.yaml:29]: assuming we just need to target packages, then I can pull down packages-all which contains both platform files.

**Comment by WilliamXieMSFT** [comment]
SBOM validation: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5848546&view=results

**Review by jander-msft — APPROVED** [review]
nit: Rename everything from "PostBuild" to "PostPackage"

**Comment by WilliamXieMSFT** [comment]
Updated SBOM validation: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5849129&view=results
